Rubberduck.VBEEditor/Events/VbeEvents.cs

```diff
@@ -29,15 +30,14 @@ public static VBEEvents Initialize(IVBE vbe)
 
         public static void Terminate()
         {
-            lock (Lock)
+            lock (_instanceLock)
             {
                 if (_instance == null)
                 {
                     return;
                 }
 
-                _instance.Dispose();
-                _instance = null;
+                _instance.TerminateInternal();
```

Comment:
Doesn't `_instance` need to be set to `null` here?

Reply:
No, because as per the comments in the PR's OP, commands are transient, so we need to persist the `Terminated` state for those that get created after the `EventsTerminated` event has been fired. Those commands would never receive the event. Furthermore, note that the `VBEEvents` is basically a singleton and is used by CW as such -- we pass in the instance to the CW during the registration, among with the `vbe` and other instances. Therefore during the host application's lifetime, there will be only one `VBEEvents` instance that will provide a consistent answer to all the commands (and in short future, other subsystems) that need to know if it's safe to proceed or not.

Reply:
The problem with that is that it breaks the class as a singleton and effectively makes it something more like a "single-use class". `Initialize` is static, and it depends on whether or not `_instance` is null - *anything* can call `Initialize(IVBE vbe)`.  If the class is _internally_ torn down in response to the `Terminate()` call it leaves it in a state where it can never be useful again until the assembly is unloaded.  If that is the intention, the lifespan and instancing should probably be re-thought entirely.

Reply:
For the record while I agree this needs to be reworked, I I think it's best in its own PR, as `IVbeEvents` isn't the only faux single-use singleton class that needs to be reworked and it would be better to handle all those fauxes together.

Reply:
Could you remind me who calls `Terminate`? (I am currently very far from an IDE.) Depending on the answer, it might be a very bad idea not to set `_instance` to `null`. More precisely, if `Terminate` gets called on every add-in shutdown, not doing it breaks reloading RD. (RD does not get unloaded on deactivation via the dialog.)

Anyway, why should not setting it to `null` should be required for commands created after termination? CW already has a fixed instance it uses completely independent of the current value of `_instance`. It only queries `Initialize` on the next load of RD.

Reply:
There are two places where it is called:

https://github.com/rubberduck-vba/Rubberduck/blob/b57c05ab52cf5a23ba084a89597117e01427aa19/Rubberduck.Parsing/VBA/RubberduckParserState.cs#L1109-L1130

https://github.com/rubberduck-vba/Rubberduck/blob/e7e27d409078fe68094d5bdb4cbdd366be6366fe/Rubberduck.VBEEditor/Events/VBENativeServices.cs#L41-L54

I need to check but IIRC, the VbeNativeService.UnhookEvents gets called first before RPS's `Dispose`. RPS's `Dispose` is mainly precautionary. 

I don't think CW's instance is independent because we inject it ourselves:
https://github.com/rubberduck-vba/Rubberduck/blob/4ded6497ce34fae399fa86bc7f0749f96cdc2204/Rubberduck.Main/Root/RubberduckIoCInstaller.cs#L843

The idea is that the `VBEEvents` needs to be linked to VBIDE's lifetime - if it's shutting down, then we need to show that. But VBIDE gives us precious little about its state besides the addin events via the `Extension` class from where we reach the `VbeNativeServices.UnhookEvents`:
https://github.com/rubberduck-vba/Rubberduck/blob/61fd569e76965342379caf2b896123739d6f0f4c/Rubberduck.Main/Extension.cs#L251-L281

As I commented to Comintern earlier, I am not happy with faux-singleton classes and I do want to fix them, but I really want to do this in its own PR. I will test again but I believe I verified that unloading & reloading works for RD.

Reply:
I'm glad that this was reviewed - I tested again and to my surprise, it did persist the state between loads, which causes the 2nd load to be non-functional. I've restored the `null` setting and revised the `_terminated` property instead, so that it is no longer single-use. This is not best as I still have in mind to improve the design of this class and other classes where their functionality are dependent on the VBIDE's lifetime.